app/actions.py

```diff
@@ -551,28 +553,56 @@ def fileLoad(self):
     else:
       self.parser = None
 
+  def restoreUserHistory(self):
+    # Restore cursor position.
+    self.cursorRow, self.cursorCol = app.history.get(['files', self.fullPath, 'cursor'], (0, 0))
+    self.view.scrollRow, self.view.scrollCol = app.history.get(['files', self.fullPath, 'scroll'], (0, 0))
+    self.penRow, self.penCol = app.history.get(['files', self.fullPath, 'pen'], (0, 0))
+    self.doSelectionMode(app.history.get(['files', self.fullPath, 'selectionMode'], 
+                         app.selectable.kSelectionNone))
+    self.markerRow, self.markerCol = app.history.get(['files', self.fullPath, 'marker'], (0, 0))
+    # Restore redo chain
+    self.redoChain = app.history.get(['files', self.fullPath, 'redoChain'], [])
+    # Restore indices
+    self.savedAtRedoIndex = app.history.get(['files', self.fullPath, 'savedAtRedoIndex'], 0)
+    self.redoIndex = self.savedAtRedoIndex
+
   def linesToData(self):
     self.data = self.doLinesToData(self.lines)
 
   def fileWrite(self):
+    # Save all user data into history
+    self.savedAtRedoIndex = self.redoIndex
+    app.history.set(
+        ['files', self.fullPath, 'redoChain'], self.redoChain)
+    app.history.set(
+        ['files', self.fullPath, 'savedAtRedoIndex'], self.savedAtRedoIndex)
     app.history.set(
         ['files', self.fullPath, 'pen'], (self.penRow, self.penCol))
+    app.history.set(
+        ['files', self.fullPath, 'cursor'], (self.cursorRow, self.cursorCol))
+    app.history.set(
+        ['files', self.fullPath, 'scroll'], (self.view.scrollRow, self.view.scrollCol))
+    app.history.set(
+        ['files', self.fullPath, 'marker'], (self.markerRow, self.markerCol))
+    app.history.set(
+        ['files', self.fullPath, 'selectionMode'], self.selectionMode)
     # Preload the message with an error that should be overwritten.
     self.setMessage('Error saving file')
     try:
       try:
-        self.stripTrailingWhiteSpace()
```

Comment:
Yeah the issue I had with this was that it created a lot of extra entries in the redo chain, so when I saved the redoChain, it didnt seem to match. I can fix this though by just saving the redo chain after stripping the white space. That might cause confusion upon reloading the file since there are now commands that the user can undo that they themselves didnt perform. What do you think about making this a shortcut instead of a setting?